token-swap/src/lib.rs

```diff
@@ -0,0 +1,863 @@
+        if *authority_info.key != Self::authority_id(program_id, swap_info.key)? {
+            return Err(Error::InvalidProgramAddress.into());
+        }
+        if *token_a_info.key != token_swap.token_a {
+            return Err(Error::InvalidInput.into());
+        }
+        if *token_b_info.key != token_swap.token_b {
+            return Err(Error::InvalidInput.into());
+        }
+        if *pool_info.key != token_swap.pool_mint {
+            return Err(Error::InvalidInput.into());
+        }
+        let token_a = Self::token_account_deserialize(token_a_info)?;
+        let token_b = Self::token_account_deserialize(token_b_info)?;
+        let pool_token = Self::token_deserialize(pool_info)?;
+
+        let invariant = Invariant {
+            token_a: token_a.amount,
+            token_b: token_b.amount,
+            fee: token_swap.fee,
+            pool: Some(pool_token.info.supply),
+        };
+
+        let (a_amount, b_amount) = invariant
+            .redeem(amount)
+            .ok_or_else(|| Error::CalculationFailure)?;
+        Self::token_transfer(
+            accounts,
+            swap_info.key,
+            authority_info.key,
+            token_a_info.key,
+            dest_token_b_info.key,
+            a_amount,
+        )?;
+        Self::token_transfer(
+            accounts,
+            swap_info.key,
+            authority_info.key,
+            token_b_info.key,
+            dest_token_a_info.key,
+            b_amount,
+        )?;
+        Self::token_burn(
+            accounts,
+            swap_info.key,
+            authority_info.key,
+            pool_info.key,
+            source_info.key,
+            amount,
+        )?;
+        Ok(())
+    }
+    /// Processes an [SwapInstruction](enum.SwapInstruction.html).
+    pub fn process(program_id: &Pubkey, accounts: &[AccountInfo], input: &[u8]) -> ProgramResult {
+        let instruction = SwapInstruction::deserialize(input)?;
+        match instruction {
+            SwapInstruction::Init(fee) => {
+                info!("Instruction: Init");
+                Self::process_init(program_id, fee, accounts)
+            }
+            SwapInstruction::Swap(amount) => {
+                info!("Instruction: Swap");
+                Self::process_swap(program_id, amount, accounts)
+            }
+            SwapInstruction::Deposit(amount) => {
+                info!("Instruction: Deposit");
+                Self::process_deposit(program_id, amount, accounts)
+            }
+            SwapInstruction::Withdraw(amount) => {
+                info!("Instruction: Withdraw");
+                Self::process_withdraw(program_id, amount, accounts)
+            }
+        }
+    }
+}
+
+// Test program id for the swap program
+#[cfg(not(target_arch = "bpf"))]
+const SWAP_PROGRAM_ID: Pubkey = Pubkey::new_from_array([2u8; 32]);
+
+/// Routes invokes to the token program, used for testing
+#[cfg(not(target_arch = "bpf"))]
+pub fn invoke_signed<'a>(
```

Comment:
ideally make this a generic SDK function that takes a processor function.

Generally I have the impression that the sdk needs a multi-program testing framework that registers program_ids (via a macro?) and automatically routes instructions if the target is not bpf. That would also solve this issue.